services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/PresignedUrlMultipartDownloaderSubscriber.java

```diff
@@ -0,0 +1,274 @@
+    private final Long configuredPartSizeInBytes;
+    private final CompletableFuture<Void> future;
+    private final Object lock = new Object();
+    private final AtomicInteger completedParts;
+    private final AtomicInteger requestsSent;
+
+    private volatile Long totalContentLength;
+    private volatile Integer totalParts;
+    private volatile String eTag;
+    private Subscription subscription;
+
+    public PresignedUrlMultipartDownloaderSubscriber(
+        S3AsyncClient s3AsyncClient,
+        PresignedUrlDownloadRequest presignedUrlDownloadRequest,
+        long configuredPartSizeInBytes) {
+        this.s3AsyncClient = s3AsyncClient;
+        this.presignedUrlDownloadRequest = presignedUrlDownloadRequest;
+        this.configuredPartSizeInBytes = configuredPartSizeInBytes;
+        this.completedParts = new AtomicInteger(0);
+        this.requestsSent = new AtomicInteger(0);
+        this.future = new CompletableFuture<>();
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (subscription != null) {
+            s.cancel();
+            return;
+        }
+        this.subscription = s;
+        s.request(1);
+    }
+
+    @Override
+    public void onNext(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
+        if (asyncResponseTransformer == null) {
+            throw new NullPointerException("onNext must not be called with null asyncResponseTransformer");
+        }
+
+        int nextPartIndex;
+        synchronized (lock) {
+            nextPartIndex = completedParts.get();
+            if (totalParts != null && nextPartIndex >= totalParts) {
+                log.debug(() -> String.format("Completing multipart download after a total of %d parts downloaded.", totalParts));
+                subscription.cancel();
+                return;
+            }
+            completedParts.incrementAndGet();
+        }
+        makeRangeRequest(nextPartIndex, asyncResponseTransformer);
+    }
+
+    private void makeRangeRequest(int partIndex,
+                                  AsyncResponseTransformer<GetObjectResponse,
+                                      GetObjectResponse> asyncResponseTransformer) {
+        PresignedUrlDownloadRequest partRequest = createRangedGetRequest(partIndex);
+        log.debug(() -> "Sending range request for part " + partIndex + " with range=" + partRequest.range());
+        
+        requestsSent.incrementAndGet();
+        s3AsyncClient.presignedUrlExtension()
+            .getObject(partRequest, asyncResponseTransformer)
+            .whenComplete((response, error) -> {
+                if (error != null) {
+                    log.debug(() -> "Error encountered during part request for part " + partIndex);
+                    handleError(error);
+                    return;
+                }
+                validatePart(response, partIndex, asyncResponseTransformer);
```

Comment:
Potential issue: even if part fails validation, we'd still call requestMoreIfNeeded (sorry I missed this previously), maybe we can return a boolean value from validatePart to indicate whether it's valid and then we can requestMoreIfNeeded only it returns true?